acme/challenge.go

```diff
@@ -446,158 +441,178 @@ func wireOIDC01Validate(ctx context.Context, ch *Challenge, db DB, jwk *jose.JSO
+}
+
+type wireDpopToken map[string]any
+
+type verifyParams struct {
+	token      string
+	key        string
+	issuer     string
+	accountJWK *jose.JSONWebKey
+	wireID     wire.ID
+	challenge  *Challenge
+	t          time.Time
+}
+
+func parseAndVerifyWireAccessToken(v verifyParams) (*wireAccessToken, *wireDpopToken, error) {
+	k, err := pemutil.Parse([]byte(v.key)) // TODO(hs): move this to earlier in the configuration process? Do it once?
 	if err != nil {
-		return storeError(ctx, db, ch, true, NewError(ErrorRejectedIdentifierType, "error finishing validation: %s", err))
+		return nil, nil, fmt.Errorf("failed parsing public key: %w", err)
 	}
 
-	// Update and store the challenge.
-	ch.Status = StatusValid
-	ch.Error = nil
-	ch.ValidatedAt = clock.Now().Format(time.RFC3339)
-
-	if err = db.UpdateChallenge(ctx, ch); err != nil {
-		return WrapErrorISE(err, "error updating challenge")
+	pk, ok := k.(ed25519.PublicKey) // TODO(hs): allow more key types
+	if !ok {
+		return nil, nil, fmt.Errorf("unexpected type: %T", k)
 	}
 
-	parsedAccessToken, err := jose.ParseSigned(wireChallengePayload.AccessToken)
+	jwt, err := jose.ParseSigned(v.token)
 	if err != nil {
-		return WrapErrorISE(err, "invalid access token")
+		return nil, nil, fmt.Errorf("failed parsing token: %w", err)
 	}
-	access := make(map[string]interface{})
-	if err := parsedAccessToken.UnsafeClaimsWithoutVerification(&access); err != nil {
-		return WrapErrorISE(err, "failed parsing access token")
+
+	var accessToken wireAccessToken
+	if err = jwt.Claims(pk, &accessToken); err != nil {
+		return nil, nil, fmt.Errorf("failed validating Wire DPoP token claims: %w", err)
 	}
 
-	rawDpop, ok := access["proof"].(string)
-	if !ok {
-		return WrapErrorISE(err, "invalid dpop proof format in access token")
+	if err := accessToken.ValidateWithLeeway(jose.Expected{
+		Time:   v.t,
+		Issuer: v.issuer,
+	}, 360*time.Second); err != nil {
+		return nil, nil, fmt.Errorf("failed validation: %w", err)
 	}
 
-	parsedDpopToken, err := jose.ParseSigned(rawDpop)
+	rawKid, err := v.accountJWK.Thumbprint(crypto.SHA256)
 	if err != nil {
-		return WrapErrorISE(err, "invalid DPoP token")
+		return nil, nil, fmt.Errorf("failed to compute JWK thumbprint")
+	}
+	kid := base64.RawURLEncoding.EncodeToString(rawKid)
+
+	if accessToken.Cnf == nil {
+		return nil, nil, errors.New("'cnf' is nil")
 	}
-	dpop := make(map[string]interface{})
-	if err := parsedDpopToken.UnsafeClaimsWithoutVerification(&dpop); err != nil {
-		return WrapErrorISE(err, "failed parsing dpop token")
+	if accessToken.Cnf.Kid != kid {
+		return nil, nil, fmt.Errorf("expected kid %q; got %q", kid, accessToken.Cnf.Kid)
+	}
+	if accessToken.ClientID != v.wireID.ClientID {
+		return nil, nil, fmt.Errorf("invalid Wire client ID %q", accessToken.ClientID)
+	}
+	if accessToken.Expiry.Time().After(v.t.Add(time.Hour * 24 * 365)) {
+		return nil, nil, fmt.Errorf("'exp' %s is too far into the future", accessToken.Expiry.Time().String())
 	}
```

Comment:
It would be weird to have tokens with a long expiration time, but why do we allow ~ one year?

Reply:
I took the value from Wire's implementation. The reason it exists is because they want to protect against clients trying to use a token that's valid for _far_ too long.